streams/src/main/java/org/apache/kafka/streams/state/Stores.java

```diff
@@ -0,0 +1,206 @@
+         * @return the key-value store
+         */
+        public <K, V> KeyValueStore<K, V> create(String name, ProcessorContext context);
+
+        /**
+         * Create a key value store that stores the given type of key and value, that records changes to a Kafka topic, and that
+         * uses the {@link #withTime(Time) configured Time} instance.
+         * 
+         * @param name the name of the store, used in the name of the topic to which entries are recorded
+         * @param context the processing context
+         * @param keyClass the class for the keys, which must be one of the types for which Kafka has built-in serializers and
+         *            deserializers (e.g., {@code String.class}, {@code Integer.class}, {@code Long.class}, or
+         *            {@code byte[].class})
+         * @param valueClass the class for the values, which must be one of the types for which Kafka has built-in serializers and
+         *            deserializers (e.g., {@code String.class}, {@code Integer.class}, {@code Long.class}, or
+         *            {@code byte[].class})
+         * @return the key-value store
+         * @throws IllegalArgumentException if any of the parameters are null, or if the {@code keyClass} or {@code valueClass}
+         *             are not one of {@code String.class},
+         *             {@code Integer.class}, {@code Long.class}, or
+         *             {@code byte[].class}
+         */
+        public <K, V> KeyValueStore<K, V> create(String name, ProcessorContext context,
+                                                 Class<K> keyClass, Class<V> valueClass);
+
+        /**
+         * Create a key value store that records changes to a Kafka topic and that uses the {@link #withTime(Time) configured
+         * Time} instance and the supplied serializers and deserializers.
+         * 
+         * @param name the name of the store, used in the name of the topic to which entries are recorded
+         * @param context the processing context
+         * @param keySerializer the serializer for keys; may not be null
+         * @param keyDeserializer the deserializer for keys; may not be null
+         * @param valueSerializer the serializer for values; may not be null
+         * @param valueDeserializer the deserializer for values; may not be null
+         * @return the key-value store
+         * @throws IllegalArgumentException if any of the parameters are null
+         */
+        public <K, V> KeyValueStore<K, V> create(String name, ProcessorContext context,
+                                                 Serializer<K> keySerializer, Deserializer<K> keyDeserializer,
+                                                 Serializer<V> valueSerializer, Deserializer<V> valueDeserializer);
+    }
+
+    /**
+     * The factory for creating in-memory key-value stores.
+     */
+    public static interface InMemoryFactory extends KeyValueStoreFactory {
```

Comment:
@guozhangwang, take a look at [this alternative](https://gist.github.com/rhauch/ad9a30bb47a041f5301d), which doe ensure type safety and works with the IDE. Requiring some definition of the key and values (rather than using the default serdes on ProcessorContext) is what makes this work. I think this addresses all of your comments. WDYT?